src/controller/java/AndroidDeviceControllerWrapper.h

```diff
@@ -0,0 +1,65 @@
+/*
+ *   Copyright (c) 2020 Project CHIP Authors
+ *   All rights reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ */
+#pragma once
+
+#include <memory>
+
+#include <jni.h>
+
+#include <controller/CHIPDeviceController.h>
+
+#include "AndroidDevicePairingDelegate.h"
+
+/**
+ * This class contains all relevant information for the JNI view of CHIPDeviceController
+ * to handle all controller-related processing.
+ *
+ * Generally it contains the DeviceController class itself, plus any related delegates/callbacks.
+ */
+class AndroidDeviceControllerWrapper
+{
+public:
+    ~AndroidDeviceControllerWrapper();
+
+    chip::DeviceController::ChipDeviceController * Controller() { return mController.get(); }
+
+    jlong ToJNIHandle()
+    {
+        static_assert(sizeof(jlong) >= sizeof(void *), "Need to store a pointer in a java handle");
+        return reinterpret_cast<jlong>(this);
+    }
+
+    static AndroidDeviceControllerWrapper * FromJNIHandle(jlong handle)
+    {
+        return reinterpret_cast<AndroidDeviceControllerWrapper *>(handle);
+    }
+
+    static AndroidDeviceControllerWrapper * AllocateNew(chip::NodeId nodeId, chip::System::Layer * systemLayer,
+                                                        chip::Inet::InetLayer * inetLayer, CHIP_ERROR * errInfoOnFailure);
+
+private:
+    using ChipDeviceControllerPtr         = std::unique_ptr<chip::DeviceController::ChipDeviceController>;
+    using AndroidDevicePairingDelegatePtr = std::unique_ptr<AndroidDevicePairingDelegate>;
+
+    ChipDeviceControllerPtr mController;
+    AndroidDevicePairingDelegatePtr mDelegate;
```

Comment:
Should we name this `mPairingDelegate` to be specific?

Reply:
Yes. Then again in my future PR I will totally clobber this since this class itself needs to run the callbacks (has access to the JNI)